src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -557,125 +613,185 @@ def _post_process(
 
     # create metadata file used by parser
-    loader_helper.create_meta_file(source_dir, assembly_upa, assembly_obj_info, genome_obj_info)
+    loader_helper.create_meta_file(source_data_dir, assembly_upa, assembly_obj_info, genome_obj_info)
 
     # create a softlink from new_dir in collectionssource to the contents of target_dir in sourcedata
     new_dir = os.path.join(collections_source_dir, assembly_upa)
-    loader_helper.create_softlink_between_dirs(new_dir, target_dir)
+    loader_helper.create_softlink_between_dirs(new_dir, ws_source_data_dir)
+
+
+def _process_batch_upload(
+        obj_tuples: list[WSObjTuple],
+        workspace_id: int,
+        load_id: str,
+        asu_client: AssemblyUtil,
+        gfu_client: GenomeFileUtil,
+        job_data_dir: str,
+        upload_assembly_only: bool = True,
+) -> list[UploadResult]:
+    if upload_assembly_only:
+        upload_results = _upload_assemblies_to_workspace(asu_client, workspace_id, load_id, obj_tuples)
+    else:
+        upload_results = _upload_genomes_to_workspace(gfu_client, workspace_id, load_id, obj_tuples, job_data_dir)
 
+    return upload_results
 
-def _upload_assembly_files_in_parallel(
-    asu: AssemblyUtil,
-    ws: Workspace,
-    upload_env_key: str,
-    workspace_id: int,
-    load_id: str,
-    collections_source_dir: str,
-    wait_to_upload_objs: dict[str, str],
-    batch_size: int,
-    source_dir: str,
+
+def _process_failed_uploads(
+        ws: Workspace,
+        workspace_id: int,
+        load_id: str,
+        obj_tuples: list[WSObjTuple],
+        upload_assembly_only: bool = True
+) -> list[UploadResult]:
+
+    # figure out uploads that succeeded
+    upload_results = list()
+    name2tuple = {obj_tuple.obj_name: obj_tuple for obj_tuple in obj_tuples}
+
+    (uploaded_assembly_obj_infos,
+     uploaded_genome_obj_infos) = _query_workspace_with_load_id_mass(ws,
+                                                                     workspace_id,
+                                                                     load_id,
+                                                                     list(name2tuple.keys()),
+                                                                     assembly_objs_only=upload_assembly_only)
+
+    if upload_assembly_only:
+        batch_uploaded_assembly_tuples = [name2tuple[info[1]] for info in uploaded_assembly_obj_infos]
+
+        for assembly_obj_info, assembly_tuple in zip(uploaded_assembly_obj_infos, batch_uploaded_assembly_tuples):
+            upload_result = UploadResult(
+                assembly_obj_info=assembly_obj_info,
+                assembly_tuple=assembly_tuple
+            )
+            upload_results.append(upload_result)
+    else:
+
+        batch_uploaded_genome_tuples = [name2tuple[info[1]] for info in uploaded_genome_obj_infos]
+        # TODO: In case of missing assembly_tuple, we need to build it.
+        for (genome_obj_info,
```

Comment:
```suggestion
        batch_uploaded_genome_tuples = [name2tuple[info[1]] for info in uploaded_genome_obj_infos]
        # TODO: In case of missing assembly_tuple, we need to build it.
        # TODO: download assembly files for successful uploads and store in standard places
        for (genome_obj_info,
```